sdk/src/transaction_context.rs

```diff
@@ -204,34 +216,70 @@ impl TransactionContext {
     }
 
     /// Used by the runtime when a new CPI instruction begins
-    pub fn record_compiled_instruction(&mut self, instruction: CompiledInstruction) {
+    pub fn record_instruction(&mut self, stack_height: usize, instruction: InstructionContext) {
         if let Some(records) = self.instruction_trace.last_mut() {
-            records.push(instruction);
+            records.push((stack_height, instruction));
         }
     }
+
+    /// Returns instruction trace
+    pub fn get_instruction_trace(&self) -> &InstructionTrace {
+        &self.instruction_trace
+    }
 }
 
+/// List of (stack height, instruction) for each top-level instruction
+pub type InstructionTrace = Vec<Vec<(usize, InstructionContext)>>;
+
 /// Loaded instruction shared between runtime and programs.
 ///
 /// This context is valid for the entire duration of a (possibly cross program) instruction being processed.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct InstructionContext {
     program_accounts: Vec<usize>,
     instruction_accounts: Vec<InstructionAccount>,
     instruction_data: Vec<u8>,
 }
 
 impl InstructionContext {
+    /// New
+    pub fn new(
+        program_accounts: &[usize],
+        instruction_accounts: &[InstructionAccount],
+        instruction_data: &[u8],
+    ) -> Self {
+        InstructionContext {
+            program_accounts: program_accounts.to_vec(),
+            instruction_accounts: instruction_accounts.to_vec(),
+            instruction_data: instruction_data.to_vec(),
+        }
+    }
+
     /// Number of program accounts
     pub fn get_number_of_program_accounts(&self) -> usize {
         self.program_accounts.len()
     }
 
+    /// Get the index of the instruction's program id
+    pub fn get_program_id_index(&self) -> usize {
+        self.program_accounts.last().cloned().unwrap_or_default()
+    }
+
+    /// Get the instruction's program id
+    pub fn get_program_id(&self, transaction_context: &TransactionContext) -> Pubkey {
+        transaction_context.account_keys[self.program_accounts.last().cloned().unwrap_or_default()]
+    }
+
     /// Number of accounts in this Instruction (without program accounts)
     pub fn get_number_of_instruction_accounts(&self) -> usize {
         self.instruction_accounts.len()
     }
 
+    /// Get the instruction's accounts
+    pub fn get_instruction_accounts(&self) -> &[InstructionAccount] {
```

Comment:
You thinking we replace it with a trace-specific instruction struct?

Reply:
I'm exposing it in the instruction trace

Reply:
We could expose an iterator of a trait that exposes the `AccountMeta` infos

Reply:
It just occurred to me that you can't know what I refer to as I haven't merged that branch yet.
I meant exposing the individual fields in `InstructionAccount` separately like this:
https://github.com/Lichtso/solana/commit/c34020df8ac1edb4ad22e2dc2ef7634ba1a6b471

Reply:
Yeah, thats what I thought you meant, but we need a way to discover the account indicies from the `InstructiionContext`.  So instead of `get_instruction_accounts()` we need a `get_instruction_account_indicies()`

Reply:
And make `InstructionAccount` private

Reply:
We already have a `get_instruction_account_indicies_in_transaction()`
it is called `InstructionContext::get_index_in_transaction()`

Reply:
Yup, way past that :-)